protobuf/types/service.proto

```diff
@@ -1,68 +1,206 @@
+    ];
+
+    // Mark a parameter as an array of the defined type.
+    bool repeated = 9 [
+      (gogoproto.moretags) = 'hash:"name:7"'
+    ];
+
+    // Optional object structure type when type is set to `Object`.
+    repeated Parameter object = 10 [
+      (gogoproto.moretags) = 'hash:"name:7" validate:"unique,dive,required"'
+    ];
   }
 
   // A configuration is the configuration of the main container of the service's instance.
   message Configuration {
-    repeated string volumes = 1;      // List of volumes.
-    repeated string volumesFrom = 2;  // List of volumes mounted from other dependencies.
-    repeated string ports = 3;        // List of ports the container exposes.
-    repeated string args = 4;         // Args to pass to the container.
-    string command = 5;               // Command to run the container.
-    repeated string env = 6;          // Default env vars to apply to service's instance on runtime.
+    // List of volumes.
+    repeated string volumes = 1 [
+      (gogoproto.moretags) = 'hash:"name:1" validate:"unique,dive,printascii"'
+    ];
+
+    // List of volumes mounted from other dependencies.
+    repeated string volumesFrom = 2 [
+      (gogoproto.moretags) = 'hash:"name:2" validate:"unique,dive,printascii"'
+    ];
+
+    // List of ports the container exposes.
+    repeated string ports = 3 [
+      (gogoproto.moretags) = 'hash:"name:3" validate:"unique,dive,portmap"'
+    ];
+
+    // Args to pass to the container.
+    repeated string args = 4 [
+      (gogoproto.moretags) = 'hash:"name:5" validate:"printascii"'
+    ];
+
+    // Command to run the container.
+    string command = 5 [
+      (gogoproto.moretags) = 'hash:"name:4" validate:"printascii"'
+    ];
+
+    // Default env vars to apply to service's instance on runtime.
+    repeated string env = 6 [
+      (gogoproto.moretags) = 'hash:"name:6" validate:"unique,dive,env"'
+    ];
   }
 
   // A dependency is a configuration of an other container that runs separately from the service.
   message Dependency {
-    string key = 8;                   // Dependency's key.
-    string image = 1;                 // Image's name of the container.
-    repeated string volumes = 2;      // List of volumes.
-    repeated string volumesFrom = 3;  // List of volumes mounted from other dependencies.
-    repeated string ports = 4;        // List of ports the container exposes.
-    repeated string args = 6;         // Args to pass to the container.
-    string command = 5;               // Command to run the container.
-    repeated string env = 9;          // Default env vars to apply to dependency on runtime.
+    // Dependency's key.
+    string key = 8 [
+      (gogoproto.moretags) = 'hash:"name:1" validate:"printascii"'
+    ];
+
+    // Image's name of the container.
+    string image = 1 [
+      (gogoproto.moretags) = 'hash:"name:2" validate:"printascii"'
+    ];
+
+    // Configuration of the container.
+    Configuration configuration = 3 [
```

Comment:
Let's find another way to do this without breaking the API ;)
One solution is to keep `Dependency` and `Configuration` as two different messages (no composition) like "before".
And then, because proto already generate the getter on the properties, a interface can be created for the functions that use both structs :)
What do you think guys?

Reply:
I think we should fix our API asap to have it clean and consistent. 

Instead of refactoring more code and creating more interfaces we should just change it to make it cleaner.

Reply:
I don't think it really cleaner.. it forces `configuration` and `dependency` to share the exact same properties and behavior. This is more an internal simplification that an external (API) system.
I'm going to do a PR from this PR to do it. I think it will be pretty simple and small.